java/memory/src/main/java/org/apache/arrow/memory/util/hash/MurmurHasher.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory.util.hash;
+
+import static io.netty.util.internal.PlatformDependent.getByte;
+import static io.netty.util.internal.PlatformDependent.getInt;
+
+import org.apache.arrow.memory.util.ArrowBufPointer;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Implementation of the Murmur hashing algorithm.
+ * Details of the algorithm can be found in
+ * https://en.wikipedia.org/wiki/MurmurHash
+ * <p>
+ *   Murmur hashing is computationally expensive, as it involves several
+ *   integer multiplications. However, the produced hash codes have
+ *   good quality in the sense that they are uniformly distributed in the universe.
+ * </p>
+ * <p>
+ *   Therefore, this algorithm is suitable for scenarios where uniform hashing
+ *   is desired (e.g. in an open addressing hash table/hash set).
+ * </p>
+ */
+public class MurmurHasher implements ArrowBufHasher {
+
+  private final int seed;
+
+  /**
+   * Creates a default Murmur hasher, with seed 0.
+   */
+  public MurmurHasher() {
+    this(0);
+  }
+
+  /**
+   * Creates a Murmur hasher.
+   * @param seed the seed for the hasher.
+   */
+  public MurmurHasher(int seed) {
+    this.seed = seed;
+  }
+
+  @Override
+  public int hashCode(long address, int length) {
+    int index = 0;
+    int hash = seed;
+    while (index + 4 <= length) {
+      int intValue = getInt(address + index);
+      if (!ArrowBufPointer.LITTLE_ENDIAN) {
```

Comment:
curious, is this required ? arrow bufs are always in little endian.

Reply:
IMO, the getInt method ultimately depends on the UNSAFE API, whose behavior is platform dependent (it can be little endian or big endian). 

Please correct me if I am wrong.